open_spiel/games/dou_dizhu/dou_dizhu_utils.cc

```diff
@@ -0,0 +1,840 @@
+      int chain_tail = chain_head + chain_length - 1;
+      num_cards = std::min(num_cards, hand[chain_tail]);
+      if(num_cards < 3) break;
+      std::vector<KickerType> all_kicker_types;
+      if(prev_action != kInvalidAction){
+        if(chain_length > prev_action_params.chain_length) break;
+        if(chain_length == prev_action_params.chain_length){
+          all_kicker_types.push_back(prev_action_params.kicker_type);
+        }
+      } else {all_kicker_types.push_back(kSolo); all_kicker_types.push_back(kPair);}
+      for(auto kicker_type: all_kicker_types){
+        std::array<int, kNumRanks> used_hand{};
+        for(int i = 0; i < chain_length; ++i) used_hand[chain_head+i] = 3;
+        dfs_add_all_airplane_kickers(chain_head, chain_length, 0, kNumRanks-1, used_hand, hand, actions, kicker_type);
+      }
+    }
+  }
+}
+
+
+
+std::array<int, kNumRanks> ActionToHand(int action){
+  std::array<int, kNumRanks> hand{};
+  if((action >= kPlayActionBase && action < kSoloChainActionBase) || 
+     (action >= kPairActionBase && action < kPairChainActionBase) || 
+     (action >= kTrioActionBase && action < kTrioWithSoloActionBase) ||
+     (action >= kBombActionBase && action < kRocketActionBase)){
+       hand = SingleRankHand(action);
+  } else if((action >= kSoloChainActionBase && action < kPairActionBase) ||
+            (action >= kPairChainActionBase && action < kTrioActionBase) ||
+            (action >= kAirplaneActionBase && action < kAirplaneWithSoloActionBase)){
+      hand = ChainOnlyHand(action);
+  } else if(action >= kTrioWithSoloActionBase && action < kAirplaneActionBase){
+    hand = SingleTrioCombHand(action);
+  } else if(action >= kAirplaneWithSoloActionBase && action < kBombActionBase){
+    hand = AirplaneCombHand(action);
+  } else if(action == kRocketActionBase){
+    hand[kNumRanks - 1] = hand[kNumRanks - 2] = 1;
+  } else {SpielFatalError("Non valid Action Ids");}
+  return hand;
+}
+
+
+void SearchForLegalActions(std::vector<Action>& legal_actions, std::array<int, kNumRanks>& hand, int prev_action){
+  if(hand[kNumRanks - 2] && hand[kNumRanks - 1]) legal_actions.push_back(kRocketActionBase);
+  if(prev_action == kInvalidAction){
+    // search for all possible actions
+    SearchSingleRankActions(legal_actions, hand, prev_action);
+    SearchChainOnlyActions(legal_actions, hand, prev_action);
+    SearchSingleTrioCombActions(legal_actions, hand, prev_action);
+    SearchAirplaneCombActions(legal_actions, hand, prev_action);
+  } else if (prev_action >= kBombActionBase && prev_action < kRocketActionBase){
+    // if previous action is a bomb, then only higher bomb or rocket can be played
+    SearchSingleRankActions(legal_actions, hand, prev_action);
+  } else {
+    // check for bombs
+    for(int rank = 0; rank < kNumRanks - 2; ++rank){
+      if(hand[rank] == kNumSuits){
+        std::array<int, kNumRanks> used_rank{};
+        used_rank[rank] = kNumSuits;
```

Comment:
std::array<int, kNumRanks> used_rank = {kNumSuits};?

Reply:
Oh in this case we do not know which rank index it is